### PR TITLE
[FLINK-17428] [table-planner-blink] supports projection push down on new table source interface in blink planner

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*supports projection push down on new table source interface in blink planner*

## Brief change log

  - *introduce PushProjectIntoTableSourceScanRule which push a projection into a table scan on new table source interface*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added PushProjectIntoTableSourceScanRuleTest to verify the plan*
  - *Extended CalcITCase to verify the result after projection*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by godfreyhe** [comment]
@flinkbot run azure

**Comment by godfreyhe** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/factories/TestProjectableValuesTableFactory.java [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+			FieldsDataType dataType = (FieldsDataType) physicalRowDataType;
+			RowType rowType = ((RowType) physicalRowDataType.getLogicalType());
+			DataTypes.Field[] fields = new DataTypes.Field[projectedFields.length];
+			for (int i = 0; i < projectedFields.length; ++i) {
+				int[] projection = projectedFields[i];
+				Preconditions.checkArgument(projection.length == 1);
+				int index = projection[0];
+				this.projectedFields[i] = index;
+				String fieldName = rowType.getFieldNames().get(index);
+				fields[i] = DataTypes.FIELD(fieldName, dataType.getFieldDataTypes().get(fieldName));
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/factories/TestProjectableValuesTableFactory.java:322]: `getFieldDataTypes` has been deleted

**Comment by KurtYoung** [comment]
still has some checkstyle error

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushProjectIntoTableSourceScanRule.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+		SupportsProjectionPushDown newProjectPushDownSource = (SupportsProjectionPushDown) newTableSource;
+
+		int[][] projectedFields = new int[usedFields.length][];
+		List<String> fieldNames = new ArrayList<>();
+		for (int i = 0; i < usedFields.length; ++i) {
+			int usedField = usedFields[i];
+			projectedFields[i] = new int[] { usedField };
+			fieldNames.add(scan.getRowType().getFieldNames().get(usedField));
+		}
+		newProjectPushDownSource.applyProjection(projectedFields);
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushProjectIntoTableSourceScanRule.java:95]:
```suggestion
		((SupportsProjectionPushDown) newTableSource).applyProjection(projectedFields);
```

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushProjectIntoTableSourceScanRule.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+		int[] usedFields = RexNodeExtractor.extractRefInputFields(project.getProjects());
+		// if no fields can be projected, we keep the original plan.
+		if (scan.getRowType().getFieldCount() == usedFields.length) {
+			return;
+		}
+
+		TableSourceTable oldTableSourceTable = scan.getTable().unwrap(TableSourceTable.class);
+		DynamicTableSource newTableSource = oldTableSourceTable.tableSource().copy();
+		SupportsProjectionPushDown newProjectPushDownSource = (SupportsProjectionPushDown) newTableSource;
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushProjectIntoTableSourceScanRule.java:86]:
```suggestion
```

**Pull request merged**